expression/builtin_json.go

```diff
@@ -51,471 +65,448 @@ var (
-		return d, errors.Trace(err)
-	}
-	return JSONSet(args, b.ctx.GetSessionVars().StmtCtx)
+func (b *builtinJSONSetSig) evalJSON(row []types.Datum) (res json.JSON, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	return jsonModify(b.args, row, json.ModifySet, sc)
 }
 
 type jsonInsertFunctionClass struct {
 	baseFunctionClass
 }
 
 type builtinJSONInsertSig struct {
-	baseBuiltinFunc
+	baseJSONBuiltinFunc
 }
 
 func (c *jsonInsertFunctionClass) getFunction(ctx context.Context, args []Expression) (builtinFunc, error) {
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinJSONInsertSig{newBaseBuiltinFunc(args, ctx)}
+	if len(args)&1 != 1 {
+		return nil, ErrIncorrectParameterCount.GenByArgs(c.funcName)
+	}
+	argTps := make([]evalTp, 0, len(args))
+	argTps = append(argTps, tpJSON)
+	for i := 1; i < len(args)-1; i += 2 {
+		argTps = append(argTps, tpString, tpJSON)
+	}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpJSON, argTps...)
+	args[0].GetType().Decimal = decimal4CastJSONDirectly
+	sig := &builtinJSONInsertSig{baseJSONBuiltinFunc{bf}}
+	sig.setPbCode(tipb.ScalarFuncSig_JsonInsertSig)
 	return sig.setSelf(sig), nil
 }
 
-func (b *builtinJSONInsertSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
-	}
-	return JSONInsert(args, b.ctx.GetSessionVars().StmtCtx)
+func (b *builtinJSONInsertSig) evalJSON(row []types.Datum) (res json.JSON, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	return jsonModify(b.args, row, json.ModifyInsert, sc)
 }
 
 type jsonReplaceFunctionClass struct {
 	baseFunctionClass
 }
 
 type builtinJSONReplaceSig struct {
-	baseBuiltinFunc
+	baseJSONBuiltinFunc
 }
 
 func (c *jsonReplaceFunctionClass) getFunction(ctx context.Context, args []Expression) (builtinFunc, error) {
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinJSONReplaceSig{newBaseBuiltinFunc(args, ctx)}
+	if len(args)&1 != 1 {
+		return nil, ErrIncorrectParameterCount.GenByArgs(c.funcName)
+	}
+	argTps := make([]evalTp, 0, len(args))
+	argTps = append(argTps, tpJSON)
+	for i := 1; i < len(args)-1; i += 2 {
+		argTps = append(argTps, tpString, tpJSON)
+	}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpJSON, argTps...)
+	args[0].GetType().Decimal = decimal4CastJSONDirectly
+	sig := &builtinJSONReplaceSig{baseJSONBuiltinFunc{bf}}
+	sig.setPbCode(tipb.ScalarFuncSig_JsonReplaceSig)
 	return sig.setSelf(sig), nil
 }
 
-func (b *builtinJSONReplaceSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
-	}
-	return JSONReplace(args, b.ctx.GetSessionVars().StmtCtx)
+func (b *builtinJSONReplaceSig) evalJSON(row []types.Datum) (res json.JSON, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	return jsonModify(b.args, row, json.ModifyReplace, sc)
```

Comment:
maybe this is a better way to record the error chain:
```go
res, isNull, err := jsonModify(b.args, row, json.ModifyReplace, sc)
return res, isNull, errors.Trace(err)
```

so do all the other `jsonModify` function calls